Add v0.2 release to README (#323)
cornerstone-data/bedrock: README.md

# bedrock

`bedrock` is a pipeline for building environmentally-extended input-output (EEIO) models and related data artifacts, like emission factors.

`bedrock` merges USEEIO and CEDA-US to build the Cornerstone U.S. model, a waypoint toward the Cornerstone global Multi-Regional EEIO model targeted for release in October 2026.
USEEIO is [a US specific legacy EEIO model](https://github.com/USEPA/USEEIO) originally developed by the US EPA.

> CEDA-US is the US portion of the [CEDA Multi-Regional EEIO model](https://openceda.org/solutions/ceda), developed by Watershed.

Reviewer: Is it right to say that it is developed by Watershed? instead of Cornerstone?

Reviewer: Just want to make sure we don't index too much on the past.

Author: It is fair to say CEDA is **developed** by Watershed, while its data products including Open CEDA are **released** through Cornerstone.

Author: we synced IRL and decided to keep the text as is.

- Several *per-flag ablation* configs each isolate a single methodological change from the baseline so the impact of each choice can be measured independently. For example, [`2025_usa_cornerstone_taxonomy.yaml`](bedrock/utils/config/configs/2025_usa_cornerstone_taxonomy.yaml) is the config for a specific choice to use Cornerstone taxonomy.

A separate `snapshot_version_or_git_sha` field specifies the baseline SHA, so diagnostic runs can compare current output against any released baseline.

### Validation and reproducibility

Snapshot tests in [`bedrock/transform/__tests__/test_usa.py`](bedrock/transform/__tests__/test_usa.py) validate the pipeline's main outputs: `B`, `Adom`, `Aimp`, `ytot`, `ydom`, `yimp`, `y_nab`, `scaled_q`, `exports`.
Each test re-derives one output from scratch and compares it to a reference parquet file in [Google Cloud Storage](https://console.cloud.google.com/storage/browser/cornerstone-default/snapshots?pageState=(%22StorageObjectListTable%22:(%22f%22:%22%255B%255D%22))), ensuring 100% numerical reproducibility.

> The tests run twice every weekday on `main` and alert the Cornerstone team on any failure or snapshot diff, guarding against refactors or upstream data updates that silently change output values in ways no unit test would catch.

Reviewer: Why twice a day but only on weekdays? Just seems weird.

Author: "Only on weekdays" assumes no active development happens over the weekend, "twice a day" accommodates our "ship frequently in small PRs" principles. Proposed change below @tdedecko

Author: ```suggestion
The tests run every workday on `main` and alert the Cornerstone team on any failure or snapshot diff, guarding against refactors or upstream data updates that silently change output values in ways no unit test would catch.
```

Author: we synced IRL and decided to keep the text as is.

# bedrock

`bedrock` is a pipeline for building environmentally-extended input-output (EEIO) models and related data artifacts, like emission factors.

> `bedrock` merges USEEIO and CEDA-US to build the Cornerstone U.S. model, a waypoint toward the Cornerstone global Multi-Regional EEIO (MRIO) model targeted for release in October 2026.

Reviewer: ```suggestion
`bedrock` merges USEEIO and CEDA-US to build the Cornerstone U.S. model, a waypoint toward the Cornerstone global Multi-Regional EEIO model targeted for release in October 2026.
```

USEEIO is [a US specific legacy EEIO model](https://github.com/USEPA/USEEIO) originally developed by the US EPA.

> CEDA-US (Comprehensive Environmental Data Archive) is the US portion of the [CEDA Multi-Regional EEIO (MRIO) model](https://openceda.org/solutions/ceda), developed by Watershed.

Reviewer: ```suggestion
CEDA-US is the US portion of the [CEDA Multi-Regional EEIO model](https://openceda.org/solutions/ceda), developed by Watershed.
```

`bedrock` brings in methodological contributions and improvements from both models and establishes a config-driven system that will be carried forward to the development of Cornerstone's global MRIO model outside of `bedrock`.

**Companion documents:**

> - **[Methodology for Cornerstone U.S. Model](https://github.com/cornerstone-data/papers/blob/v0.2model/us-methods/us-methods.md)** covers the equations, data sources, and methodology choices adopted from USEEIO vs CEDA.

Reviewer: This link is a reference to main branch. Should we reference a specific gitsha in the case that underlying us-methods.md gets moved over time?

Same for the link below.

> - **[Cornerstone Technical Architecture Vision](https://github.com/cornerstone-data/papers/blob/published/architecture_vision/Cornerstone_Architecture_Vision.md)** describes the Extract-Transform-Load (ETL) monorepo architecture adopted in `bedrock`.

Reviewer: ```suggestion
- **[Cornerstone Technical Architecture Vision](https://github.com/cornerstone-data/papers/blob/published/architecture_vision/Cornerstone_Architecture_Vision.md)** describes the monorepo architecture adopted in `bedrock`.
```

## Methodology choices captured in configuration

Every methodological decision involved in merging USEEIO and CEDA-US is governed by the configuration system defined in [`USAConfig`](bedrock/utils/config/usa_config.py).
Each flag in the configuration represents a discrete methodology choice, and changes from the baseline can be validated in isolation.

These flags are grouped into themes:

| Theme | Example flags |
|---|---|
| Schema / taxonomy | `use_cornerstone_2026_model_schema`, `implement_waste_disaggregation` |
| Economic IOT (input-output tables) | `use_E_data_year_for_x_in_B`, `iot_before_or_after_redefinition` |
| GHG attribution — sector methods | `update_transportation_ghg_method`, `update_electricity_ghg_method` |
| GHG attribution — gas coverage | `update_flowsa_refrigerant_method`, `add_new_ghg_activities` |
| Data vintage | `model_base_year`, `usa_base_io_data_year`, `ipcc_ar_version` |

See [`USAConfig`](bedrock/utils/config/usa_config.py) for the full list.

### Configuration files

All configuration files are in [`bedrock/utils/config/configs/`](bedrock/utils/config/configs/), where:

> - A single *full-model* config represents a full set of methodology choices made for a release. For example, `2025_usa_cornerstone_full_model.yaml` is now the default config in `get_usa_config()`.

Reviewer: ```suggestion
- A single *full-model* config represents a full set of methodology choices made for a data release. For example, `2025_usa_cornerstone_full_model.yaml` is now the default config in `get_usa_config()`.
```

Reviewer: Please link the yaml directly in the github repo.

> - A single *full-model* config represents a full set of methodology choices made for a release. For example, `2025_usa_cornerstone_full_model.yaml` is the config for the `bedrock` v0.2 release, now the default in `get_usa_config()`.

Reviewer: @MoLi7 Everything in the repo is the current state of the world for the repo. I don't believe we need to cite the release specifically in the repo at all. You don't know need to say anything about v0.2. You should just talk about it as if this is the current state of the world. Remember this documentation is included in the release itself.

Reviewer: My recommendation is to remove all references to the version in the readme.

Author: Thanks @tdedecko , they're removed in [8560661](https://github.com/cornerstone-data/bedrock/pull/323/commits/8560661952e965b29ec67389d4c23b99eeb2a496).

> - Several *per-flag ablation* configs each isolate a single methodological change from the baseline so the impact of each choice can be measured independently. For example, `2025_usa_cornerstone_taxonomy.yaml` is the config for a specific choice to use Cornerstone taxonomy.

Reviewer: Please link the yaml directly in the github repo.

A separate `snapshot_version_or_git_sha` field specifies the baseline SHA, so diagnostic runs can compare current output against any released baseline.

### Validation and reproducibility

Snapshot tests in [`bedrock/transform/__tests__/test_usa.py`](bedrock/transform/__tests__/test_usa.py) validate the pipeline's main outputs: `B`, `Adom`, `Aimp`, `ytot`, `ydom`, `yimp`, `y_nab`, `scaled_q`, `exports`.

> Each test re-derives one output from scratch and compares it to a reference parquet file in [Google Cloud Storage](https://console.cloud.google.com/storage/browser/cornerstone-default/snapshots?pageState=(%22StorageObjectListTable%22:(%22f%22:%22%255B%255D%22))) (GCS), ensuring 100% numerical reproducibility.

Reviewer: ```suggestion
Each test re-derives one output from scratch and compares it to a reference parquet file in [Google Cloud Storage](https://console.cloud.google.com/storage/browser/cornerstone-default/snapshots?pageState=(%22StorageObjectListTable%22:(%22f%22:%22%255B%255D%22))), ensuring 100% numerical reproducibility.
```

The tests run twice every weekday on `main` and alert the Cornerstone team on any failure or snapshot diff, guarding against refactors or upstream data updates that silently change output values in ways no unit test would catch.

> Snapshots are stored in GCS, one folder per git SHA, and `bedrock/utils/snapshots/.SNAPSHOT_KEY` pins the SHA currently under test.

Reviewer: `bedrock/utils/snapshots/.SNAPSHOT_KEY` make this an explicit link to the file.

When methodology changes, snapshots are regenerated and uploaded under the new SHA, and the pin is bumped.
Old snapshots are retained, so every prior release remains independently reproducible.

> `uv.lock` pins Python and all dependencies, so the exact environment for any snapshot or historical result can be reconstructed.

Reviewer: `uv.lock` make this an explicit link to the file.

Also is this needed to call out? This seems like a environment setup context sort of thing. Consider removing this line.

Author: deleted since we have an installation guidance in a later paragraph

### Diagnostics of model results compared to USEEIO and CEDA-US baselines

> A work-in-progress diagnostics report comparing `bedrock` outputs against USEEIO and CEDA-US can be found in [Cornerstone's papers repository](https://github.com/cornerstone-data/papers/pull/12).

Reviewer: This is the reference to a pull request. Can we reference a specific file?

`bedrock` brings in methodological contributions and improvements from both models and establishes a config-driven system that will be carried forward to the development of Cornerstone's global MRIO model outside of `bedrock`.

**Companion documents:**
- **[Methodology for Cornerstone U.S. Model](https://github.com/cornerstone-data/papers/blob/f2372a0d72def7a9961279ab22982adac07e5f67/us-methods/us-methods.md)** covers the equations, data sources, and methodology choices adopted from USEEIO vs CEDA.

> - **[Cornerstone Technical Architecture Vision](https://github.com/cornerstone-data/papers/blob/449b7981b537be419dc861b49fac559547b09bae/architecture_vision/Cornerstone_Architecture_Vision.md)** describes the monorepo architecture adopted in `bedrock`.

Author: @tdedecko Here I'm using gitsha instead of branch name

### Configuration files

All configuration files are in [`bedrock/utils/config/configs/`](bedrock/utils/config/configs/), where:
- A single *full-model* config represents a full set of methodology choices made for a data release. For example, [`2025_usa_cornerstone_full_model.yaml`](bedrock/utils/config/configs/2025_usa_cornerstone_full_model.yaml) is now the default config in `get_usa_config()`.

> - Several *per-flag ablation* configs each isolate a single methodological change from the baseline so the impact of each choice can be measured independently. For example, [`2025_usa_cornerstone_taxonomy.yaml`](bedrock/utils/config/configs/2025_usa_cornerstone_taxonomy.yaml) is the config for a specific choice to use Cornerstone taxonomy.

Author: @tdedecko URL embedded

Each test re-derives one output from scratch and compares it to a reference parquet file in [Google Cloud Storage](https://console.cloud.google.com/storage/browser/cornerstone-default/snapshots?pageState=(%22StorageObjectListTable%22:(%22f%22:%22%255B%255D%22))), ensuring 100% numerical reproducibility.
The tests run twice every weekday on `main` and alert the Cornerstone team on any failure or snapshot diff, guarding against refactors or upstream data updates that silently change output values in ways no unit test would catch.

> Snapshots are stored in GCS, one folder per git SHA, and [`bedrock/utils/snapshots/.SNAPSHOT_KEY`](bedrock/utils/snapshots/.SNAPSHOT_KEY) pins the SHA currently under test.

Author: @tdedecko URL embedded

When methodology changes, snapshots are regenerated and uploaded under the new SHA, and the pin is bumped.
Old snapshots are retained, so every prior release remains independently reproducible.

### Diagnostics of model results compared to USEEIO and CEDA-US baselines

> A work-in-progress diagnostics report comparing `bedrock` outputs against USEEIO and CEDA-US can be found in [here](https://github.com/cornerstone-data/papers/blob/diagnostics-report-v0-2/diagnostics_report/bedrock_v0_2_to_v0_diagnostics.md) in Cornerstone's papers repository.

Author: @tdedecko linking to the WIP file on the branch (not using gitsha to avoid 404 error).
